### ISSUE TITLE
Processing "Union" wrong result

### ISSUE THREAD

**Issue by gioman** [qgis/QGIS#42145]
QGIS 3.18.0

Take two polygon layers, with some overlap between them but **not** within them, the Union results are as expected > there output features were cut along **all** boundaries of the input layers

[union1.zip](https://github.com/qgis/QGIS/files/6109583/union1.zip)

![Screenshot_20210309_151412](https://user-images.githubusercontent.com/1951107/110495205-31bc6700-80ec-11eb-8acf-b03ea6c863a3.png)

Now take a similar example, but where there are also overlaps **within** (at least) one of the input layers: the result is not correct as some output feature where not cut along all boundaries, and in particular the ones within each input overlap

[union2.zip](https://github.com/qgis/QGIS/files/6109598/union2.zip)

![Screenshot_20210309_151825](https://user-images.githubusercontent.com/1951107/110495519-7c3de380-80ec-11eb-9671-2031315adb64.png)

**Comment by uclaros** [comment]
Your example is a little confusing, but I think that layer A is cut with layer B boundaries and layer B is cut with layer A boundaries, which is what I would expect.

**Comment by agiudiceandrea** [comment]
From the Union tool help, the first sentence is clear about how the tool works with an Overlay layer:
> An Overlay layer can also be used, in which case features from each layer are split at their overlap with features from the other one,

maybe the second sentence could be a little bit misleading

> creating a layer containing all the portions from both input and overlay layers

Anyway the example image is clear:

![image](https://user-images.githubusercontent.com/16253859/110545461-6dc3ec00-812d-11eb-84d7-6b4e37cfbe95.png)

**Comment by gioman** [comment]
> Anyway the example image is clear:

@agiudiceandrea source of the image?

**Comment by agiudiceandrea** [comment]
@gioman sorry: the Union tool help in the user manual https://docs.qgis.org/3.16/en/docs/user_manual/processing_algs/qgis/vectoroverlay.html#union

The "note" is more esplicit:
> Note
> For union(A,B) algorithm, if there are overlaps among geometries of layer A or among geometries of layer B, these are not resolved: you need to do union(union(A,B)) to resolve all overlaps, i.e. run single layer union(X) on the produced result X=union(A,B).

**Comment by gioman** [comment]
> @gioman sorry: the Union tool help in the user manual https://docs.qgis.org/3.16/en/docs/user_manual/processing_algs/qgis/vectoroverlay.html#union
> 
> The "note" is more explicit:
> 
> > Note
> > For union(A,B) algorithm, if there are overlaps among geometries of layer A or among geometries of layer B, these are not resolved: you need to do union(union(A,B)) to resolve all overlaps, i.e. run single layer union(X) on the produced result X=union(A,B).

@agiudiceandrea so the results are coherent with docs, which is good. What I'm arguing here is that the results are wrong, as there should be a cut also where 1 and 2 overlap. I'm pretty sure (but not 100% sure as I don't have an ArcGIS copy at hand) that ESRI does that. This allows among other things to use the union tool using twice the same layer as input, and in this way "isolate" all the overlapping areas > something that to be achieved tight now in QGIS requires a  multi step, cumbersome process.

**Comment by agiudiceandrea** [comment]
Yes, the Union tool in ESRI ArcMap/ArcGIS works differently: it doesn't have an Input Layer and an Overlay Layer as inputs but it processes one or more Input Layers [[1](http://webhelp.esri.com/arcgisdesktop/9.3/index.cfm?TopicName=How%20Union%20(Analysis)%20works)] [[2](https://pro.arcgis.com/en/pro-app/latest/tool-reference/analysis/how-union-analysis-works.htm)].

Maybe these PR https://github.com/qgis/QGIS/pull/6859 https://github.com/qgis/QGIS/pull/6961 could shed more light on the design choices.

**Comment by gioman** [comment]
> Maybe these PR #6859 #6961 could shed more light on the design choices.

the correct result seems to me the one shown in the image in the description of https://github.com/qgis/QGIS/pull/6961#issue-187071684

If that PR was about implementing it then is not working as expected.

**Comment by gioman** [comment]
> it processes one or more Input Layers [[1](http://webhelp.esri.com/arcgisdesktop/9.3/index.cfm?TopicName=How%20Union%20(Analysis)%20works)] [[2](https://pro.arcgis.com/en/pro-app/latest/tool-reference/analysis/how-union-analysis-works.htm)]

@agiudiceandrea QGIS be able to union/intersects multiple layers and not just 2 would be very good indeed (and In fact I believe there is a decade old feature request about it), but anyway the focus here is on another aspect > How is should result and union/intersection among 2 overlapping layers when there are overlaps also within the input layers.

**Comment by agiudiceandrea** [comment]
> the correct result seems to me the one shown in the image in the description of [#6961 (comment)](https://github.com/qgis/QGIS/pull/6961#issue-187071684)

That image, as you can also see in the user manual, is about the behaviour of the tool with only the Input Layer. The other image is about the behaviour of the tool with an Input Layer and an Overlay Layer.

In https://github.com/qgis/QGIS/pull/6961/commits/64b8c7250c06ad41b0aaf4193bbd1e584e7f0b12 @wonder-sk more explicitly explains that:

> So now we have two variants of union:
> - union(A) - does union within geometries of one layer
> - union(A,B) - does union between geometries of two layers
> 
> For union(A,B) algorithm if there are overlaps among geometries of layer A or among geometries of layer B,
> these are not resolved: one needs to do union(union(A,B)) to resolve all overlaps, i.e. run single layer
> union(X) on the produced result X=union(A,B)

**Comment by agiudiceandrea** [comment]
It seems the QGIS Union tools behave like the SAGA 2.3.2 Polygon Union tool (unchecking the Split Parts option), while the ArcMap/ArcGIS Union tool like the GRASS 7.8.5 v.overlay (with the "or" operator, except that ArcMap creates multipart features and GRASS creates singlepart features).

**Comment by gioman** [comment]
> It seems the QGIS Union tools behaves like the SAGA 2.3.2 Polygon Union tool (unchecking the Split Parts option), while the ArcMap/ArcGIS Union tool like the GRASS 7.8.5 v.overlay (with the "or" operator, except that ArcMap creates multipart features and GRASS creates singlepart features

@agiudiceandrea I have to review yet your previous comments, but anyway this is the point... most of users expect it to behave like v.overlay and how works in AG.

**Comment by gioman** [comment]
> This means that, to get to the final result pretended by @gioman you now need to pass table A and table B as single layers in a previous union operation.

@SrNetoChan exactly. It is needed to do two unions using just 1 input layer, and then union the results. Problem is that frequently the ovelaps **within** a layer may not be macroscopic, so a user could just miss to do the first step...



> An option to preprocess the input layers would be a nice option, for keep both behaviors available.

agree. I think that also an option to "split parts" is necessary (as SAGA has) because all the above has a good chance of creating multi-part geometries.


I could see this become a feature request, but honestly I think is a bug.

**Comment by SrNetoChan** [comment]
For now, IMHO, it's a feature request.

Maybe a crowd funding to add those options and allowing multiple inputs could work. This is a very used functionality.

**Comment by gioman** [comment]
> For now, IMHO, it's a feature request.

@SrNetoChan ok, you are the boss ;)

**Comment by agiudiceandrea** [comment]
> It is needed to do two unions using just 1 input layer, and then union the results.

Anyway, with two layers, only two Union are needed, not three: union(union(A,B)) to obtain the same result as ArcMap/ArcGIS

**Comment by gioman** [comment]
> Anyway, with two layers, only two Union are needed union(union(A,B)), not three, to obtain the same result as ArcMap/ArcGIS

@agiudiceandrea correct, still one too many ;)
